main-command/src/main/scala/sbt/internal/ConsoleChannel.scala

```diff
@@ -23,13 +25,19 @@ private[sbt] final class ConsoleChannel(val name: String) extends CommandChannel
       case Some(pf) => pf(s)
       case None     => "> "
     }
-    val reader = new FullReader(history, s.combinedParser, JLine.HandleCONT, true)
+    val reader =
+      new FullReader(history, s.combinedParser, JLine.HandleCONT, Terminal.throwOnClosedSystemIn)
     override def run(): Unit = {
       // This internally handles thread interruption and returns Some("")
-      val line = reader.readLine(prompt)
-      line match {
-        case Some(cmd) => append(Exec(cmd, Some(Exec.newExecId), Some(CommandSource(name))))
-        case None      => append(Exec("exit", Some(Exec.newExecId), Some(CommandSource(name))))
+      try {
+        reader.readLine(prompt) match {
+          case Some(cmd) => append(Exec(cmd, Some(Exec.newExecId), Some(CommandSource(name))))
+          case None =>
+            println("") // Prevents server shutdown log lines from appearing on the prompt line
+            append(Exec("exit", Some(Exec.newExecId), Some(CommandSource(name))))
+        }
+      } catch {
+        case _: ClosedChannelException =>
```

Comment:
What would be a situation where the input would close itself?

Reply:
This is the purpose of `Terminal.throwOnClosedSystemIn`. The problem is that `ConsoleReader.readLine` returns null both if the user inputs EOF via <ctrl+d> and if the call to read returns -1. In that latter case, we don't want to shutdown sbt because we may be running scripted or a server session. Before the wrapped system in overrode read to poll until `System.in.available > 0`. This kept JLine from returning null when `System.in` wasn't attached.